How many late days are you using for this assignment? 

In [63]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler

# Part 1: Load the dataset

In [64]:
# Load the given datasets
ckd_num = pd.read_csv("data/chronic_kidney_disease_numerical.csv")
ckd_cat = pd.read_csv("data/chronic_kidney_disease_categorical.csv")

# Print the data
print(ckd_num)
print(ckd_cat)


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0   5.7  
3     8200.0   4.5  
4     

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
Each data is a numerical or categorical data (nominal) that describes the biological data of the patient. The features and the labels are the variables. Categorical features: red blood cell, hypertension, anemia. Numerical features: Age, blood pressure, sugar. 


# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [65]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
ckd_num.unique_id.duplicated()

# Drop these duplicate rows
ckd_num.loc[~ckd_num.unique_id.duplicated(), :]

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
ckd_cat.unique_id.duplicated()
ckd_cat.loc[~ckd_cat.unique_id.duplicated(), :]



,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
396,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
398,492993,3.0,0.0,NaN,normal,notpresent,notpresent,yes,no,no,good,no,yes,ckd
399,273038,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

All the unique ids are not present in both datasets. When we combine them, there are indexes that have no input (NaN) for either their categorical variables or numerical ones. The rows missing data from on of the datasets have NaN for the columns that initially belonged to the other dataset in the combined table. 

In [66]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_ckd = pd.merge(ckd_cat, ckd_num, on='unique_id', how='inner')

#Print the combined dataset
print(merged_ckd)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
2       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
3       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
4       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
477     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
478     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
479     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   
480     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   
481     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm cad  ...    bp    bgr    bu   sc    sod  pot  hemo   pcv     wbcc 

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [67]:
# Calculate the percentage of rows that contain atleast one missing value
missing_val_rows = merged_ckd.isnull().any(axis=1).sum()
percentage_missing = (missing_val_rows/ len(merged_ckd)) * 100

# Print %
print(percentage_missing)
# Drop these rows from the dataset
merged_ckd_full = merged_ckd.dropna()

# Print the Dataset
print(merged_ckd_full)

58.921161825726145
     unique_id   al   su       rbc      pc         pcc          ba htn   dm  \
5       881763  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
6       114717  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
8       901528  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
9       433613  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
10      621332  1.0  0.0  abnormal  normal  notpresent  notpresent  no  yes   
..         ...  ...  ...       ...     ...         ...         ...  ..  ...   
468     261137  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
469     431871  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
470     431871  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
471     769512  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
472     769512  0.0  0.0    normal  normal  notpresent  notpresent  no   no   

    cad  ...    bp    bgr    bu 

## Part 3.4: Sort the dataset according to the Labels

In [68]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
merged_ckd_sorted = merged_ckd_full.sort_values(by='Target').reset_index(drop=True)
# Print the dataset
print(merged_ckd_sorted)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       129053  4.0  0.0  abnormal  abnormal  notpresent     present   no   
1       124304  3.0  4.0    normal  abnormal  notpresent  notpresent  yes   
2       234299  4.0  0.0    normal  abnormal  notpresent  notpresent  yes   
3       843362  4.0  0.0    normal    normal  notpresent  notpresent  yes   
4       160040  3.0  0.0  abnormal    normal     present  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
193     963946  0.0  0.0    normal    normal  notpresent  notpresent   no   
194     203694  0.0  0.0    normal    normal  notpresent  notpresent   no   
195     471614  0.0  0.0    normal    normal  notpresent  notpresent   no   
196     411659  0.0  0.0    normal    normal  notpresent  notpresent   no   
197     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm  cad  ...     bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0 

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [69]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

merged_ckd_cat = pd.get_dummies(merged_ckd_sorted, columns=categorical_columns)

# Print the dataset
print(merged_ckd_cat)

     unique_id   al   su   age     bp    bgr     bu    sc    sod  pot  ...  \
0       129053  4.0  0.0  48.0   80.0  133.0  139.0   8.5  132.0  5.5  ...   
1       124304  3.0  4.0  69.0   70.0  264.0   87.0   2.7  130.0  4.0  ...   
2       234299  4.0  0.0  48.0   80.0   95.0  163.0   7.7  136.0  3.8  ...   
3       843362  4.0  0.0  71.0   60.0  118.0  125.0   5.3  136.0  4.9  ...   
4       160040  3.0  0.0  48.0  110.0  106.0  215.0  15.2  120.0  5.7  ...   
..         ...  ...  ...   ...    ...    ...    ...   ...    ...  ...  ...   
193     963946  0.0  0.0  50.0   80.0  137.0   46.0   0.8  139.0  5.0  ...   
194     203694  0.0  0.0  38.0   80.0   99.0   19.0   0.5  147.0  3.5  ...   
195     471614  0.0  0.0  52.0   80.0  128.0   30.0   1.2  140.0  4.5  ...   
196     411659  0.0  0.0  32.0   60.0  102.0   17.0   0.4  147.0  4.7  ...   
197     769512  0.0  0.0  38.0   60.0   91.0   36.0   0.7  135.0  3.7  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  an

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: 

If we have a variable that takes in more than 2 values, we would have to have different numerical categories such as 0, 1, 2, 3, 4. But by doing so we can wrongly suggest that this is a numerical variable and not categorical. 

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [70]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

#standard deviation
std = merged_ckd_cat[numerical_columns].std()
#bounds
mean = merged_ckd_cat[numerical_columns].mean()
upper_bound = mean + (3 * std)
lower_bound = mean - (3 * std)

#remove the outliers
mask = merged_ckd_cat[numerical_columns].apply(lambda x: (x >= lower_bound[x.name]) & (x <= upper_bound[x.name]), axis=0)

merged_ckd_cat_cleaned = merged_ckd_cat[mask.all(axis=1)]
# Print the dataset
print(merged_ckd_cat_cleaned)

     unique_id   al   su   age    bp    bgr     bu   sc    sod  pot  ...  \
0       129053  4.0  0.0  48.0  80.0  133.0  139.0  8.5  132.0  5.5  ...   
1       124304  3.0  4.0  69.0  70.0  264.0   87.0  2.7  130.0  4.0  ...   
2       234299  4.0  0.0  48.0  80.0   95.0  163.0  7.7  136.0  3.8  ...   
3       843362  4.0  0.0  71.0  60.0  118.0  125.0  5.3  136.0  4.9  ...   
7       343710  2.0  0.0  61.0  80.0  173.0  148.0  3.9  135.0  5.2  ...   
..         ...  ...  ...   ...   ...    ...    ...  ...    ...  ...  ...   
193     963946  0.0  0.0  50.0  80.0  137.0   46.0  0.8  139.0  5.0  ...   
194     203694  0.0  0.0  38.0  80.0   99.0   19.0  0.5  147.0  3.5  ...   
195     471614  0.0  0.0  52.0  80.0  128.0   30.0  1.2  140.0  4.5  ...   
196     411659  0.0  0.0  32.0  60.0  102.0   17.0  0.4  147.0  4.7  ...   
197     769512  0.0  0.0  38.0  60.0   91.0   36.0  0.7  135.0  3.7  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
0      

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [71]:
# Normalize the all Numerical Attributes in the dataset.

numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
merged_ckd_min_max = merged_ckd_cat_cleaned.copy()

scaler = MinMaxScaler()

merged_ckd_min_max[numerical_columns] = scaler.fit_transform(merged_ckd_cat_cleaned[numerical_columns])
# Print the dataset
print(merged_ckd_min_max)

     unique_id   al   su       age    bp       bgr        bu        sc  \
0       129053  4.0  0.0  0.545455  0.50  0.270386  0.843137  1.000000   
1       124304  3.0  4.0  0.818182  0.25  0.832618  0.503268  0.283951   
2       234299  4.0  0.0  0.545455  0.50  0.107296  1.000000  0.901235   
3       843362  4.0  0.0  0.844156  0.00  0.206009  0.751634  0.604938   
7       343710  2.0  0.0  0.714286  0.50  0.442060  0.901961  0.432099   
..         ...  ...  ...       ...   ...       ...       ...       ...   
193     963946  0.0  0.0  0.571429  0.50  0.287554  0.235294  0.049383   
194     203694  0.0  0.0  0.415584  0.50  0.124464  0.058824  0.012346   
195     471614  0.0  0.0  0.597403  0.50  0.248927  0.130719  0.098765   
196     411659  0.0  0.0  0.337662  0.00  0.137339  0.045752  0.000000   
197     769512  0.0  0.0  0.415584  0.00  0.090129  0.169935  0.037037   

          sod       pot  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0    0.485714  0.896552  ...   

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

Yes, the unique_id is not appropriate for modeling and predictions as the id is not related to the biological data of whether a patient has ckd or not. 

In [ ]:
#Remove that column
cleaned_ckd = merged_ckd_min_max.drop(labels='unique_id', axis=1)
#Print the dataset
print(cleaned_ckd)

      al   su       age    bp       bgr        bu        sc       sod  \
0    4.0  0.0  0.545455  0.50  0.270386  0.843137  1.000000  0.485714   
1    3.0  4.0  0.818182  0.25  0.832618  0.503268  0.283951  0.428571   
2    4.0  0.0  0.545455  0.50  0.107296  1.000000  0.901235  0.600000   
3    4.0  0.0  0.844156  0.00  0.206009  0.751634  0.604938  0.600000   
7    2.0  0.0  0.714286  0.50  0.442060  0.901961  0.432099  0.571429   
..   ...  ...       ...   ...       ...       ...       ...       ...   
193  0.0  0.0  0.571429  0.50  0.287554  0.235294  0.049383  0.685714   
194  0.0  0.0  0.415584  0.50  0.124464  0.058824  0.012346  0.914286   
195  0.0  0.0  0.597403  0.50  0.248927  0.130719  0.098765  0.714286   
196  0.0  0.0  0.337662  0.00  0.137339  0.045752  0.000000  0.914286   
197  0.0  0.0  0.415584  0.00  0.090129  0.169935  0.037037  0.571429   

          pot      hemo  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0    0.896552  0.257426  ...    True    Fal

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [73]:
# Export the dataframe to a new csv file
cleaned_ckd.to_csv("cleaned_ckd_data.csv", index = False)
# Export the dataframe to a new json file
cleaned_ckd.to_json("cleaned_ckd_data.json", orient="records")

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [76]:
cleaned_ckd.head(15)

,albumin,sugar,age,blood pressure,blood glucose random,blood urea,serum creatinine,sodium,potassium,hemoglobin,...,coronary artery disease no,coronary artery disease yes,appet_good,appet_poor,pedal edema no,pedal edema yes,anemia no,anemia yes,Target_ckd,Target_notckd
0,4.0,0.0,0.545455,0.50,0.270386,0.843137,1.000000,0.485714,0.896552,0.257426,...,True,False,True,False,False,True,True,False,True,False
1,3.0,4.0,0.818182,0.25,0.832618,0.503268,0.283951,0.428571,0.379310,0.475248,...,False,True,True,False,False,True,True,False,True,False
2,4.0,0.0,0.545455,0.50,0.107296,1.000000,0.901235,0.600000,0.310345,0.207921,...,True,False,True,False,True,False,False,True,True,False
3,4.0,0.0,0.844156,0.00,0.206009,0.751634,0.604938,0.600000,0.689655,0.366337,...,True,False,False,True,False,True,True,False,True,False
7,2.0,0.0,0.714286,0.50,0.442060,0.901961,0.432099,0.571429,0.793103,0.000000,...,False,True,False,True,False,True,False,True,True,False
11,4.0,3.0,0.818182,0.25,0.618026,0.562092,0.728395,0.142857,0.344828,0.168317,...,False,True,True,False,False,True,False,True,True,False
12,4.0,2.0,0.753247,1.00,0.399142,0.287582,0.839506,0.714286,0.586207,0.019802,...,True,False,True,False,False,True,True,False,True,False
16,2.0,0.0,0.649351,0.75,0.253219,0.633987,0.777778,0.457143,0.655172,0.138614,...,True,False,True,False,True,False,True,False,True,False
18,3.0,0.0,0.844156,0.25,0.639485,0.470588,0.395062,0.514286,0.517241,0.267327,...,False,True,True,False,True,False,True,False,True,False
20,4.0,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.571429,0.689655,0.217822,...,True,False,False,True,True,False,True,False,True,False


Paste here: 

|   | albumin | sugar | age     | blood pressure | blood glucose random | blood urea | serum creatinine | sodium  | potassium | hemoglobin | ... | coronary artery disease no | coronary artery disease yes | appet_good | appet_poor | pedal edema no | pedal edema yes | anemia no | anemia yes | Target_ckd | Target_notckd |
|---|---------|-------|---------|----------------|----------------------|------------|------------------|---------|-----------|------------|-----|----------------------------|----------------------------|------------|------------|----------------|----------------|------------|------------|------------|--------------|
| 0 | 4.0     | 0.0   | 0.545455| 0.50           | 0.270386             | 0.843137   | 1.000000         | 0.485714 | 0.896552  | 0.257426   | ... | True                       | False                      | True       | False      | False          | True           | True       | False      | True       | False        |
| 1 | 3.0     | 4.0   | 0.818182| 0.25           | 0.832618             | 0.503268   | 0.283951         | 0.428571 | 0.379310  | 0.475248   | ... | False                      | True                       | True       | False      | False          | True           | True       | False      | True       | False        |
| 2 | 4.0     | 0.0   | 0.545455| 0.50           | 0.107296             | 1.000000   | 0.901235         | 0.600000 | 0.310345  | 0.207921   | ... | True                       | False                      | True       | False      | True           | False          | False      | True       | True       | False        |
| 3 | 4.0     | 0.0   | 0.844156| 0.00           | 0.206009             | 0.751634   | 0.604938         | 0.600000 | 0.689655  | 0.366337   | ... | True                       | False                      | False      | True       | False          | True           | True       | False      | True       | False        |
| 7 | 2.0     | 0.0   | 0.714286| 0.50           | 0.442060             | 0.901961   | 0.432099         | 0.571429 | 0.793103  | 0.000000   | ... | False                      | True                       | False      | True       | False          | True           | False      | True       | True       | False        |
| 11 | 4.0    | 3.0   | 0.818182| 0.25           | 0.618026             | 0.562092   | 0.728395         | 0.142857 | 0.344828  | 0.168317   | ... | False                      | True                       | True       | False      | False          | True           | False      | True       | True       | False        |
| 12 | 4.0    | 2.0   | 0.753247| 1.00           | 0.399142             | 0.287582   | 0.839506         | 0.714286 | 0.586207  | 0.019802   | ... | True                       | False                      | True       | False      | False          | True           | True       | False      | True       | False        |
| 16 | 2.0    | 0.0   | 0.649351| 0.75           | 0.253219             | 0.633987   | 0.777778         | 0.457143 | 0.655172  | 0.138614   | ... | True                       | False                      | True       | False      | True           | False          | True       | False      | True       | False        |
| 18 | 3.0    | 0.0   | 0.844156| 0.25           | 0.639485             | 0.470588   | 0.395062         | 0.514286 | 0.517241  | 0.267327   | ... | False                      | True                       | True       | False      | True           | False          | True       | False      | True       | False        |
| 20 | 4.0    | 0.0   | 0.000000| 0.00           | 0.103004             | 0.372549   | 0.074074         | 0.571429 | 0.689655  | 0.217822   | ... | True                       | False                      | False      | True       | True           | False          | True       | False      | True       | False        |
| 21 | 4.0    | 1.0   | 0.649351| 0.25           | 0.600858             | 0.104575   | 0.160494         | 0.600000 | 0.310345  | 0.831683   | ... | True                       | False                      | True       | False      | True           | False          | True       | False      | True       | False        |
| 22 | 4.0    | 1.0   | 0.753247| 0.00           | 0.725322             | 0.313725   | 0.481481         | 0.628571 | 0.862069  | 0.178218   | ... | True                       | False                      | False      | True       | False          | True           | True       | False      | True       | False        |
| 23 | 4.0    | 0.0   | 0.194805| 0.75           | 0.158798             | 0.196078   | 0.160494         | 0.285714 | 0.206897  | 0.059406   | ... | True                       | False                      | True       | False      | True           | False          | False      | True       | True       | False        |
| 24 | 3.0    | 0.0   | 1.000000| 0.25           | 0.137339             | 0.326797   | 0.271605         | 0.000000 | 0.965517  | 0.099010   | ... | True                       | False                      | False      | True       | True           | False          | False      | True       | True       | False        |
| 26 | 3.0    | 2.0   | 0.870130| 1.00           | 0.965665             | 0.522876   | 0.641975         | 0.714286 | 0.000000  | 0.148515   | ... | False                      | True                       | False      | True       | True           | False          | True       | False      | True       | False        |


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer: I could mitigate data conversion hallucinations by inputing data in batches. 

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [74]:
# Code to rename all the columns in the dataset
# Dictionary for renaming the columns
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
rename_dict = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc_normal': 'red blood cells normal',
    'rbc_abnormal': 'red blood cells abnormal',
    'pc_normal': 'pus cell normal',
    'pc_abnormal': 'pus cell abnormal',
    'pcc_present': 'pus cell clumps present',
    'pcc_notpresent': 'pus cell clumps notpresent',
    'ba_present': 'bacteria present',
    'ba_notpresent': 'bactera notpresent',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wbcc': 'white blood cell count',
    'rbcc': 'red blood cell count',
    'htn_yes': 'hypertension yes',
    'htn_no': 'hypertension no',
    'dm_yes': 'diabetes mellitus yes',
    'dm_no': 'diabetes mellitus no',
    'cad_yes': 'coronary artery disease yes',
    'cad_no': 'coronary artery disease no',
    'appet_yes': 'appetite yes',
    'appet_no': 'appetite no',
    'pe_yes': 'pedal edema yes',
    'pe_no': 'pedal edema no',
    'ane_yes': 'anemia yes',
    'ane_no': 'anemia no',
}

# Renaming the columns
cleaned_ckd.rename(columns=rename_dict, inplace=True)

# Display the updated DataFrame
print(cleaned_ckd)

     albumin  sugar       age  blood pressure  blood glucose random  \
0        4.0    0.0  0.545455            0.50              0.270386   
1        3.0    4.0  0.818182            0.25              0.832618   
2        4.0    0.0  0.545455            0.50              0.107296   
3        4.0    0.0  0.844156            0.00              0.206009   
7        2.0    0.0  0.714286            0.50              0.442060   
..       ...    ...       ...             ...                   ...   
193      0.0    0.0  0.571429            0.50              0.287554   
194      0.0    0.0  0.415584            0.50              0.124464   
195      0.0    0.0  0.597403            0.50              0.248927   
196      0.0    0.0  0.337662            0.00              0.137339   
197      0.0    0.0  0.415584            0.00              0.090129   

     blood urea  serum creatinine    sodium  potassium  hemoglobin  ...  \
0      0.843137          1.000000  0.485714   0.896552    0.257426  ... 

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [75]:
# Converted SQL to Pandas code
result = cleaned_ckd.groupby('Target_ckd').size().reset_index(name='count')
print(result)

   Target_ckd  count
0       False    150
1        True     23


*Describe what the above code does here*

I think that because we are only querying the Target_ckd column, the above code outputs the number of indexes in the final clean data which have been classified as having ckd only. The False and True shows us how much of indexes that have been classified as ckd are actually those with ckd (true) or those who were falsely classified (false). 